proposals/p3763.md

```diff
@@ -0,0 +1,783 @@
+declarations in the `interface`:
+
+```carbon
+interface I {
+  fn F(s: Self);
+}
+impl i32 as I {
+  fn F(s: i32);
+}
+```
+
+It is tempting to base the rules here on the rules we use to check
+redeclarations. However, this turns out to be a poor choice:
+
+-   The redeclaration rule would syntactically couple the declaration in the
+    `interface` to the definition in the `impl`. But these declarations could be
+    in different libraries, or even different packages, so such coupling would
+    make refactorings that change the way that code is expressed but not its
+    meaning either difficult or impossible.
+-   The associated function in an `impl` is expected to have different syntax
+    than that in the interface in some cases. The two declarations are in
+    different scopes, so will refer to the same types in different ways. And the
+    declaration in the `impl` is declared with knowledge of the `Self` type and
+    associated constants for the interface, which it may be reasonable to use
+    directly in the declaration of the function.
+
+Therefore, a different rule is used. In the general case, when matching an
+associated function in an `impl` against a function declaration in an
+`interface`, a synthetic function called a _thunk_ is generated:
+
+-   The declaration of the thunk is formed by substituting the `Self` type of
+    the `impl` into the declaration in the `interface`. This implicitly also
+    provides values for any associated constants used in the declaration.
+-   The body of the thunk calls the function in the `impl`, passing in the
+    arguments to the thunk, and, if a return type is specified in the interface,
+    returning the value returned by the call.
+
+If the function in the interface does not have a return type, the program is
+invalid if the function in the `impl` specifies a return type.
+
+> **Note:** Another rule might work better here. For example, we could allow
+> `impl`s to add a return type in anticipation of the `interface` later adding
+> one. However, such a feature is considered out of scope for this proposal.
+
+Implicit conversions are performed as necessary to initialize the parameters of
+the function in the `impl` from the parameters of the thunk, and to initialize
+the return value of the function from the result of the call.
+
+As a special case, if a `var` parameter in the callee is initialized from a
+`var` parameter of the thunk of the same type, no copy is performed, and the
+parameters refer to the same variable. In this case, the `var` parameter is
+destroyed by either the caller of the thunk or the callee, depending on the
+calling convention, and not by the thunk itself.
+
+As an optimization, if the parameter types and return type are the same between
+the thunk and its callee, and the calling convention is otherwise the same, the
+implementation should use the callee function directly and not generate a thunk.
```

Comment:
These special cases worry me a bit about the approach.

Is it possible to frame the semantics here to constructively avoid the need for this? The implicit conversions thing is something we already get when calling a function -- is there a way to avoid the indirection through a thunk when framing this?

Reply:
Maybe there's some nice way to do that, but I've not found it yet.

I think there end up somewhat fundamentally being two separate implicit conversion steps here. For example, given:

```carbon
interface A {
  fn F(a: i16);
}
class C {}
impl C as A {
  fn F(a: i32) { Print("F"); }
}
fn G(T:! A, a: i8) { T.F(a); }
fn H(a: i8) { G(C, a); }
```

... I think we first want to convert from `i8` to `i16` in the call to `T.F`, and then from `i16` to `i32` in the `impl`. I think that's going to be necessary because in general we'll be type-checking the two halves of this conversion separately. And in the general case there might not be an implicit conversion directly from whichever type is in the place of `i8` to whichever type is in the place of `i32`, or the combined conversion might do something different from the two individual conversions (though the latter case concerns me less than the former).

In the above example, I *think* I also want to reject `(C as A).F(n)` given `n: i32` -- even though the function in the `impl` supports an `i32` argument, I think `(C as A).F` should implement `Call(like i16)`, like `T` does. But that seems less clear-cut to me: we know more about `C as A` than we do about `T`, so perhaps it's OK for us to also know more about `(C as A).F` than about `T.F` -- but I think it'd be problematic if it *only* implements `Call(like i32)` and not `Call(like i16)`, because that'd mean that a checked generic call that works before monomorphization doesn't work afterwards / if the call is manually monomorphized (if the argument is a type that implicitly converts to `i16` but not to `i32`). I think this works out better if `(C as A).F` is something like "thunk for `F` with the signature of `A.F`" rather than simply being `F`.

However, one potentially nice observation is that these special cases aren't actually special just for thunks. We get the exact same special cases in the converting `impl` of `Call` for an `fn` as we get in the converting `impl` of `Call` for a thunk. In both cases, we get an implicit-conversion `impl` of `Call` that doesn't copy `var`s, and an exact-match `impl` of `Call` that doesn't recursively use `Call` again. Maybe once we understand better how `var` interacts with the `Call` interface, the special case for `var` might even disappear here.

Reply:
I agree that we'll want a thunk in the general case, and I like that.

My suggestion was that it would be nice to have a preferred mode where there is no thunk at all that is tried first rather than framing eliding the thunk as an optimization. Specifically, I'm imagining we say that if the types match exactly between impl and interface, we do the non-thunk thing, but if there is any difference, we then build a thunk. And I wonder if we want to preclude `var` parameters when the thunk is required?

If it's important to build from the thunk up, we could talk about notionally generating a thunk, and then if the call is an *exact* match, we call the `impl`  function dircetly rather than the thunk. And again, maybe only then is `var` allowed?

Either way, I think it would be good to make the text here a bit more clear that the `impl` declaration *doesn`t* redeclare the `interface` declaration in any way, that it is a new thing, but that these declarations <TERM> (correspond?) through the notional thunk building?

FWIW, also happy to just leave the `var` behavior as a future work thing, it doesn't seem important to get right here.

Reply:
Done.